python/cudf/cudf/core/indexed_frame.py

```diff
@@ -414,3 +415,55 @@ def sort_index(
         if ignore_index is True:
             out = out.reset_index(drop=True)
         return self._mimic_inplace(out, inplace=inplace)
+
+    def _gather(self, gather_map, keep_index=True, nullify=False):
+        gather_map = cudf.core.column.as_column(gather_map)
+        if not is_integer_dtype(gather_map.dtype):
+            gather_map = gather_map.astype("int32")
+
+        if not _gather_map_is_valid(gather_map, len(self), nullify):
+            raise IndexError("Gather map index is out of bounds.")
+
+        result = self.__class__._from_maybe_indexed_columns(
+            libcudf.copying.gather(
+                list(self._index._columns + self._columns)
+                if keep_index
+                else list(self._columns),
+                gather_map,
+                nullify=nullify,
+            ),
+            self._column_names,
+            self._index.names if keep_index else None,
+        )
+
+        result._copy_type_metadata(self, include_index=keep_index)
+        if keep_index:
+            result.index.names = self._index.names
+        return result
+
+    def _positions_from_column_names(self, column_names, include_index=False):
+        """Map each column name into their positions in the frame.
+
+        Set `include_index=True` to add the number of index columns to the
+        result.
+        """
+        n_indices_cols = len(self._index._data) if include_index else 0
+        return [
+            i + n_indices_cols
+            for i, name in enumerate(self._column_names)
+            if name in column_names
+        ]
```

Comment:
This seems like it could be refactored like in previous suggestions.

Reply:
ditto, `_column_name_to_indices` is removed in favor of this.